File: HW2/Q2 - Coding/q8_e.m
import brml.*
load("dataset.dat")
load("joint.dat")

assignments = zeros(12, 1);
givens = [8 11]; % input givens
givenAssignments = [0 0 0 0 0 0 0 0 1 0 0 1]; % hasSummer is the left-most node
queries = [1]; % input query variable, and add 1 to it because we're working with 1-based indexing
for g = givens % preprocessing, because we're using 1-based indexing
    assignments(g + 1) = 1;
end
assignments = assignments + 1;
for i = 1:length(queries) % preprocessing, because we're using 1-based indexing
    %queries(i, :) = queries(i, :) + 1;
end

newSymptoms = zeros(7, 16);
newDiseases = zeros(4, 2);

for i = 5:11
    if any(givens == i) == true % when we marginalize over a symptom, its distribution becomes 1
        newSymptoms(i-4, :) = symptoms(i-4, :, assignments(i+1)); % since the variable is given, we take the column indexed for 'True', which is column 2
    elseif any(queries == i) == false
        newSymptoms(i-4, :) = ones(16, 1);
    end
end

messageTable = ones(16, 1); % parameterized by 4 binary variables 2^4 = 16 states
firstMarginalizedDisease = 1;
for i = 4:-1:1 % find the first disease to marginalize over to initialize our message table
    if any(givens == i) == false && any(queries == i) == false % marginalize in order from 1, 2, 3, and 4, and find the first one we marginalize over
        firstMarginalizedDisease = i;
        break
    end
end
numDiseasesNotMarginalized = 4; % there are 4 diseases, and this dictates the size of messageTable at each step
variablesLeft = [1 2 3 4];
hasSummerFalseVarFalse = diseases(firstMarginalizedDisease, 1, 1);
hasSummerFalseVarTrue = diseases(firstMarginalizedDisease, 1, 2);
hasSummerTrueVarFalse = diseases(firstMarginalizedDisease, 2, 1);
hasSummerTrueVarTrue = diseases(firstMarginalizedDisease, 2, 2);
% marginalize out the probability distributions of this disease
for j = 0:(cast(length(messageTable)/2, 'int64')-1) % TODO: switch j to go from 1 to length(messageTable)-1    
    %fprintf("here, iteration %d\n", j);
    %fprintf("marginalizing node: %d\n", firstMarginalizedDisease);
    assign = de2bi(j, numDiseasesNotMarginalized);
    newAssign = zeros(4, 1); %assign(setdiff(1:numDiseasesNotMarginalized, firstMarginalizedDisease));
    currAssignmentIndex = 1;
    for k = 1:4
        if any(givens == k) == false
            newAssign(k) = assign(currAssignmentIndex);
            currAssignmentIndex = currAssignmentIndex + 1;
        end
    end
    
    falseAssign = newAssign.';
    falseAssign(firstMarginalizedDisease) = 0;
    falseAssignDecimal = bi2de(falseAssign);
    trueAssign = newAssign.';
    trueAssign(firstMarginalizedDisease) = 1;
    trueAssignDecimal = bi2de(trueAssign);
    %disp(falseAssign);
    %disp(trueAssign);
    fprintf("here2 %d %d\n", falseAssignDecimal, trueAssignDecimal);
    
    probFalse = 1.0;
    probTrue = 1.0;
    for givenIndex = givens 
        if givenIndex >= 5 % if the given is a leaf node
            fprintf("given: %d, val: %f\n", givenIndex, newSymptoms(givenIndex-4, trueAssignDecimal+1));
            probFalse = probFalse * newSymptoms(givenIndex-4, falseAssignDecimal+1); % the marginalized variable is false
            probTrue = probTrue * newSymptoms(givenIndex-4, trueAssignDecimal+1); % else, the marginalized variable is true
        end
    end

    %fprintf("%f %f %f %f\n", hasSummerFalseVarFalse, hasSummerFalseVarTrue, hasSummerTrueVarFalse, hasSummerTrueVarTrue);
    %fprintf("%f %f\n", probFalse, probTrue);
    varFalseNew = hasSummerFalseVarFalse * probFalse + hasSummerFalseVarTrue * probTrue;
    varTrueNew = hasSummerTrueVarFalse * probFalse + hasSummerTrueVarTrue * probTrue;
    %fprintf("%f %f\n", varFalseNew, varTrueNew);

    varFalseNewIndex = bi2de([0 assign]);
    messageTable(varFalseNewIndex+1, :) = varFalseNew;

    varTrueNewIndex = bi2de([1 assign]);
    messageTable(varTrueNewIndex+1, :) = varTrueNew;
    %disp([0 assign]);
    %disp([1 assign]);
    %fprintf("%d %d\n", varFalseNewIndex, varTrueNewIndex);
    fprintf("messageTable(%d, :) = %f -- %f %f %f %f\n", varFalseNewIndex+1, varFalseNew, hasSummerFalseVarFalse, probFalse, hasSummerFalseVarTrue, probTrue);
    fprintf("messageTable(%d, :) = %f -- %f %f %f %f\n", varTrueNewIndex+1, varTrueNew, hasSummerTrueVarFalse, probFalse, hasSummerTrueVarTrue, probTrue);
end
numDiseasesNotMarginalized = numDiseasesNotMarginalized - 1;
variablesLeft(find(variablesLeft == firstMarginalizedDisease)) = [];
variablesLeft = [0 variablesLeft];
%error("END OF CODE BLOCK")
%END THE ABOVE CODE

for i = (firstMarginalizedDisease-1):-1:1 %(firstMarginalizedDisease+1)
    if any(givens == i) == true % given variable
        newDiseases(i, :) = diseases(i, :, 2);
    elseif any(queries == i) == false % not a query or given variable, so marginalize over it
        fprintf("marginalizing node %d\n", i);
        newMessageTable = zeros(2^numDiseasesNotMarginalized, 1);
        
        hasSummerFalseVarFalse = diseases(i, 1, 1);
        hasSummerFalseVarTrue = diseases(i, 1, 2);
        hasSummerTrueVarFalse = diseases(i, 2, 1);
        hasSummerTrueVarTrue = diseases(i, 2, 2);
        % marginalize out the probability distributions of this disease
        %{
        for j = 0:(cast(length(messageTable)/2, 'int64')-1)
            assign = de2bi(j, 4);
            newAssign = assign(setdiff(1:4, i));

            falseAssign = assign;
            falseAssign(numDiseasesNotMarginalized) = 0;
            falseAssignDecimal = bi2de(falseAssign);
            trueAssign = assign;
            trueAssign(numDiseasesNotMarginalized) = 1;
            trueAssignDecimal = bi2de(trueAssign);

            probFalse = messageTable(falseAssignDecimal+1); % the marginalized variable is false
            probTrue = messageTable(trueAssignDecimal+1); % else, the marginalized variable is true

            varFalseNew = hasSummerFalseVarFalse * probFalse + hasSummerFalseVarTrue * probTrue;
            varTrueNew = hasSummerTrueVarFalse * probFalse + hasSummerTrueVarTrue * probTrue;

            varFalseNewIndex = bi2de([0 newAssign]);
            newMessageTable(varFalseNewIndex+1, :) = varFalseNew;

            varTruenewIndex = bi2de([1 newAssign]);
            newMessageTable(varTrueNewIndex+1, :) = varTrueNew;
        end
        %}
        
        for j = 0:(cast(length(newMessageTable), 'int64')-1) % TODO: switch j to go from 1 to length(messageTable)-1
            %{
            %fprintf("here, iteration %d\n", j);
            %fprintf("marginalizing node: %d\n", firstMarginalizedDisease);
            assign = de2bi(j, 4);
            %newAssign = zeros(4, 1); %assign(setdiff(1:numDiseasesNotMarginalized, firstMarginalizedDisease));
            %currAssignmentIndex = 1;
            %for k = 1:4
            %    if any(givens == k) == false
            %        newAssign(k) = assign(currAssignmentIndex);
            %        currAssignmentIndex = currAssignmentIndex + 1;
            %    end
            %end
            newAssign = [0 assign.'];

            falseAssign = newAssign.';
            falseAssign(i) = 0;
            falseAssignDecimal = bi2de(falseAssign);
            trueAssign = newAssign.';
            trueAssign(i) = 1;
            trueAssignDecimal = bi2de(trueAssign);
            %disp(falseAssign);
            %disp(trueAssign);
            %fprintf("here2 %d %d\n", falseAssignDecimal, trueAssignDecimal);
            
            falseRetrieveMessage = bi2de([newAssign 0]);
            trueRetrieveMessage = bi2de([newAssign 1]);
            probFalse = messageTable(falseRetrieveMessage+1);
            probTrue = messageTable(trueRetrieveMessage+1);

            %fprintf("%f %f %f %f\n", hasSummerFalseVarFalse, hasSummerFalseVarTrue, hasSummerTrueVarFalse, hasSummerTrueVarTrue);
            %fprintf("%f %f\n", probFalse, probTrue);
            varFalseNew = hasSummerFalseVarFalse * probFalse + hasSummerFalseVarTrue * probTrue;
            varTrueNew = hasSummerTrueVarFalse * probFalse + hasSummerTrueVarTrue * probTrue;
            %fprintf("%f %f\n", varFalseNew, varTrueNew);

            varFalseNewIndex = bi2de([0 assign]);
            messageTable(varFalseNewIndex+1, :) = varFalseNew;

            varTrueNewIndex = bi2de([1 assign]);
            messageTable(varTrueNewIndex+1, :) = varTrueNew;
            disp([0 assign]);
            disp([1 assign]);
            %fprintf("%d %d\n", varFalseNewIndex, varTrueNewIndex);
            %}
            
            assign = de2bi(j, numDiseasesNotMarginalized);
            
            falseAssign = [assign(1:end) 0];
            falseAssignDecimal = bi2de(falseAssign);
            trueAssign = [assign(1:end) 1];
            disp("here");
            disp(assign);
            disp(falseAssign);
            disp(trueAssign);
            trueAssignDecimal = bi2de(trueAssign);
            probFalse = messageTable(falseAssignDecimal + 1);
            probTrue = messageTable(trueAssignDecimal + 1);
            
            varFalseNew = diseases(i, assign(1)+1, 1) * probFalse;
            varTrueNew = diseases(i, assign(1)+1, 2) * probTrue;
            
            newMessageTable(j+1, :) = varFalseNew + varTrueNew;
        end
        
        messageTable = newMessageTable;
        numDiseasesNotMarginalized = numDiseasesNotMarginalized - 1;
    
        if i == 2
            error("MARGINALIZING OVER NODE 2")
        elsif
            disp("MARGINALIZING OVER ANOTHER NODE")
        end
    end
end

% sum over hasSummer
% TODO: hardcoded this, change it later
finalTable = zeros(2, 1);
finalTable(1) = summer_probs(1)*diseases(1, 1, 1)*messageTable(bi2de([0 0])+1) + summer_probs(2)*diseases(1, 2, 1)*messageTable(bi2de([1 0])+1);
finalTable(2) = summer_probs(1)*diseases(1, 1, 2)*messageTable(bi2de([0 1])+1) + summer_probs(2)*diseases(1, 2, 2)*messageTable(bi2de([1 1])+1);

disp(finalTable(1))
disp(finalTable(2))
fprintf("Prob x_1 == FALSE: %f\n", finalTable(1) / (finalTable(1) + finalTable(2)));
fprintf("Prob x_1 == TRUE: %f\n", finalTable(2) / (finalTable(1) + finalTable(2)));

File: HW2/Q2 - Coding/q8_e.py
import numpy as np
from scipy.io import loadmat

class Factor():
    def __init__(self, table, condVars, inputVars, givenVars, givenVals, symptomsInit = False, diseasesInit = False, summerInit = False):
        self.table = table
        self.condVars = condVars
        self.inputVars = inputVars
        self.allVars = self.condVars + self.inputVars

        diseaseIndices = [1, 2, 3, 4]
        self.diseasesLeft = [index for index in diseaseIndices if not index in givenVars]

        self.convert_matlab_table()

        if symptomsInit:
            # only keep the dimensions that are not given
            if diseaseIndices[0] in givenVars:
                givenVarIndex = givenVars.index(diseaseIndices[0])
                self.table = self.table[givenVals[givenVarIndex]]
                self.condVars.remove(diseaseIndices[0])
            if diseaseIndices[1] in givenVars:
                givenVarIndex = givenVars.index(diseaseIndices[1])
                self.table = self.table[:, givenVals[givenVarIndex]]
                self.condVars.remove(diseaseIndices[1])
            if diseaseIndices[2] in givenVars:
                givenVarIndex = givenVars.index(diseaseIndices[2])
                self.table = self.table[:, :, givenVals[givenVarIndex]]
                self.condVars.remove(diseaseIndices[2])
            if diseaseIndices[3] in givenVars:
                givenVarIndex = givenVars.index(diseaseIndices[3])
                self.table = self.table[:, :, :, givenVals[givenVarIndex]]
                self.condVars.remove(diseaseIndices[3])
            if inputVars[0] in givenVars:
                givenVarIndex = givenVars.index(inputVars[0])
                self.table = self.table[:, :, :, :, givenVals[givenVarIndex]]
                self.inputVars.remove(inputVars[0])
        elif diseasesInit:
            # only keep the dimensions that are not given
            if 0 in givenVars:
                givenVarIndex = givenVars.index(0)
                self.table = self.table[givenVals[givenVarIndex]]
                self.condVars.remove(0)
            if inputVars[0] in givenVars:
                givenVarIndex = givenVars.index(inputVars[0])
                self.table = self.table[:, givenVals[givenVarIndex]]
                self.inputVars.remove(inputVars[0])
        elif summerInit:
            if 0 in givenVars:
                self.table = np.array([self.table[givenVals[0]]])
                self.condVars.remove(0)

        self.allVars = self.condVars + self.inputVars
        self.table = self.table.squeeze()

    def mult_factor_symptom(self, newFactor):
        messageTable = self.table
        newTable = newFactor.get_table()

        if len(self.table.shape) == len(self.diseasesLeft) and len(newTable.shape) == len(self.diseasesLeft):
            newMessageTable = messageTable
            newVars = self.diseasesLeft
        else:
            newMessageTable = np.stack((messageTable, messageTable), axis = -1) # converts from 2x2x2x2x2 to 2x2x2x2x2x2 (1 more dimension)
            newVars = list(self.diseasesLeft)
            newVars += self.allVars if not len(self.table.shape) == len(self.diseasesLeft) else []
            newVars += newFactor.allVars if not len(newTable.shape) == len(self.diseasesLeft) else []
        newVars = list(set(newVars))

        for i in range(0, int(2**len(newVars))):
            assignments = toBinaryArray(i, len(newVars)) #len(self.diseasesLeft)) # currently storing assignments for x1, x2, x3, x4, and x10
            diseaseAssignments = assignments[0:len(self.diseasesLeft)] # currently storing assignments for x1, x2, x3, x4

            currTable = newTable
            for assignment in diseaseAssignments: # traverse the table to get this combination's assignment
                currTable = currTable[assignment] # newTable[x1][x2][x3][x4]
            newFalseTrueVars = []
            if len(newTable.shape) == len(self.diseasesLeft): # this table is a table with a given variable (e.g. x8), so it doesn't have a true and false column, just a true column (its shape is 2x2x2x2, not 2x2x2x2x2)
                newFalseTrueVars.append(currTable)
                newFalseTrueVars.append(currTable)
            else:
                newFalseTrueVars.append(currTable[0]) # x10 = false
                newFalseTrueVars.append(currTable[1]) # x10 = true

            currNewMessageTable = newMessageTable
            for assignment in assignments[:-2]:
                currNewMessageTable = currNewMessageTable[assignment]
            if len(self.table.shape) == len(self.diseasesLeft):
                currNewMessageTable = currNewMessageTable[assignments[-2]]
                currNewMessageTable[assignments[-1]] *= newFalseTrueVars[0]
            else:
                # TODO: control for the case when we have a 2x2x2x2x1 and a 2x2x2x2x2 table
                # multiplying a 2x2x2x2x2 table by a 2x2x2x2x2 table
                #currNewMessageTable = currNewMessageTable[assignments[-1]]
                #print(i, currNewMessageTable.shape)
                for val in range(0, len(currNewMessageTable)): # is it 1 or 2?
                    currNewMessageTable[val][0] *= newFalseTrueVars[0]
                    currNewMessageTable[val][1] *= newFalseTrueVars[1]

        # handle some logistics
        self.condVars = newVars
        self.inputVars = []
        self.allVars = newVars
        self.table = newMessageTable.squeeze()
        assert(len(self.allVars) == len(self.table.shape))

    def mult_factor_disease(self, newFactor, marginalizedVar):
        newVars = set(self.get_vars()).union(set(newFactor.get_vars())) # combine [0, 4] and [1, 2, 3, 4] into set(0, 1, 2, 3)
        newVars.remove(marginalizedVar)
        newVars = sorted(list(newVars)) # convert from set(0, 1, 2, 3) to [0, 1, 2, 3]

        newFactorTable = np.zeros((2, ) * len(newVars))
        for i in range(0, 2**(len(newVars))): # from 0 to 7 inclusive
            assignments = toBinaryArray(i, len(newVars)) # each variable in this assignment corresponds to the respective variable in newVars
            varAssignments = {}
            for index, assignment in enumerate(assignments):
                varAssignments[newVars[index]] = assignment # place all variable assignments into a collective dict

            marginalizedVarFalseAssignment = dict(varAssignments) # create variable assignment for when the marginalized variable is false
            marginalizedVarTrueAssignment = dict(varAssignments) # create variable assignment for when the marginalized variable is true
            marginalizedVarFalseAssignment[marginalizedVar] = 0
            marginalizedVarTrueAssignment[marginalizedVar] = 1

            newFactorFalse = newFactor.getProb(marginalizedVarFalseAssignment) # disease table probability when marginalized variable is false
            newFactorTrue = newFactor.getProb(marginalizedVarTrueAssignment) # disease table prob when marginalized variable is true
            thisFalse = self.getProb(marginalizedVarFalseAssignment) # messageTable probability when marginalized variable is false
            thisTrue = self.getProb(marginalizedVarTrueAssignment) # messageTable probability when marginalized variable is true

            newProb = newFactorFalse * thisFalse + newFactorTrue * thisTrue # sum over the marginalized variable, and for each summation, there is one multiplication

            # traverse the new factor table and insert this newly calculated probability
            currTable = newFactorTable
            for assignment in assignments[:-1]:
                currTable = currTable[assignment]
            currTable[assignments[-1]] = newProb

        # handle some higher-level logistics
        self.condVars = newVars
        self.inputVars = []
        self.allVars = newVars
        self.table = newFactorTable.squeeze()
        assert(len(self.allVars) == len(self.table.shape))

    def mult_factor_multiple(self, newFactors, marginalizedVar):
        if len(newFactors) == 1: # special case, only happens when the query is P(x0)
            self.condVars = newFactors[0].get_vars()
            self.inputVars = []
            self.allVars = self.condVars
            self.table = newFactors[0].get_table()

        newVars = set(newFactors[0].get_vars())
        for factor in newFactors:
            newVars = newVars.union(factor.get_vars())
        if marginalizedVar in newVars:
            newVars.remove(marginalizedVar)
        newVars = sorted(list(newVars))

        newFactorTable = np.zeros((2, ) * len(newVars))
        for i in range(2**len(newVars)):
            assignments = toBinaryArray(i, len(newVars))
            varAssignments = {}
            for index, assignment in enumerate(assignments):
                varAssignments[newVars[index]] = assignment # place all variable assignments into a collective dict

            marginalizedVarFalseAssignment = dict(varAssignments) # create variable assignment for when the marginalized variable is false
            marginalizedVarTrueAssignment = dict(varAssignments) # create variable assignment for when the marginalized variable is true
            marginalizedVarFalseAssignment[marginalizedVar] = 0
            marginalizedVarTrueAssignment[marginalizedVar] = 1

            newProbFalse = 1.0
            newProbTrue = 1.0
            for factor in newFactors:
                factorFalse = factor.getProb(marginalizedVarFalseAssignment) # disease table probability when marginalized variable is false
                factorTrue = factor.getProb(marginalizedVarTrueAssignment) # disease table prob when marginalized variable is true

                newProbFalse *= factorFalse
                newProbTrue *= factorTrue

            newProb = newProbFalse + newProbTrue # sum over the marginalized variable, and for each summation, there is one multiplication

            # traverse the new factor table and insert this newly calculated probability
            currTable = newFactorTable
            for assignment in assignments[:-1]:
                currTable = currTable[assignment]
            currTable[assignments[-1]] = newProb

        # handle some higher-level logistics
        self.condVars = newVars
        self.inputVars = []
        self.allVars = newVars
        self.table = newFactorTable.squeeze()
        assert(len(self.allVars) == len(self.table.shape))

    def getProb(self, varAssignments):
        if self.table.size == 1:
            return self.table

        assignments = []
        for var in self.allVars: # if vars = [1, 2, 3, 4], then map varAssignments {4: 1, 0: 1, 2: 0, 3: 1, 1: 0} to [0, 0, 1, 1]
            assignments.append(varAssignments[var])

        # iterate over the table and find the probability that matches the corresponding variable assignments
        currTable = self.table
        for assignment in assignments:
            currTable = currTable[assignment]

        return currTable

    def convert_matlab_table(self):
        newTable = np.zeros((2, ) * (len(self.allVars))) # convert the table from 16x2 to 2x2x2x2x2

        if len(self.allVars) == 1: # special case: isSummer table
            newTable[0] = self.table[0]
            newTable[1] = self.table[1]
            self.table = newTable
            return

        for i in range(0, newTable.size): # loop through all the possible 16 states
            assignments = toBinaryArray(i, len(newTable.shape)) # convert from i = 5 to [0, 1, 1]

            currTable = newTable
            for assignment in assignments[:-1]: # traverse to currTable[0][1]
                currTable = currTable[assignment]
            condAssignmentToDecimal = toDecimal(assignments[:-1]) # convert [x1, x2, x3, x4] to the appropriate decimal value
            currTable[assignments[-1]] = self.table[condAssignmentToDecimal][assignments[-1]] # assign currTable[0][1][1] = self.table[5]
        self.table = newTable

    def get_table(self):
        return self.table

    def get_vars(self):
        return self.allVars

    def print_table(self):
        for i in range(0, self.table.size):
            binary = toBinaryArray(i, len(self.table.shape))
            currTable = self.table
            for assignment in binary:
                currTable = currTable[assignment]
            print(binary, currTable)

    def normalize_table(self):
        sum_probs = np.sum(self.table)
        self.table /= sum_probs

def toBinaryArray(number, length = None):
    if length is not None:
        return [int(ch) for ch in list(format(number, 'b').zfill(length)[::-1])] # all elements are ints e.g. [1, 0, 1, 0, 0, 0, 0, 0] for int(5) -- LSB is at index 0
    else:
        return [int(ch) for ch in list(format(number, 'b')[::-1])]

def toDecimal(bitArray):
    val = 0
    for index, bit in enumerate(bitArray):
        val += bit * (2**index)
    return int(val)

assignments = np.zeros((12, 1))
givens = [] # user input -- the nodes that are given
givenVals = [] # user input -- binary values, the assignments of the nodes in the givens list
queries = [] # user input -- the nodes we are querying
which_query = 1
if which_query == 1:
    givens = [8, 11]
    givenVals = [1, 1]
    queries = [1]
elif which_query == 2:
    givens = [4]
    givenVals = [1]
    queries = [7, 8, 9, 10, 11]
elif which_query == 3:
    givens = [0]
    givenVals = [1]
    queries = [10]
diseasesIndices = [1, 2, 3, 4]
for g in givens:
    assignments[g] = 1
data = loadmat("q8_vars.mat")
symptoms = data["symptoms"]
diseases = data["diseases"]
summerTable = data["summer_probs"]

variablesLeft = [1, 2, 3, 4]
messageTable = None

remainingFactors = []

# combine the symptoms tables
for i in range(5, 12): # 5 to 11 inclusive
    if not i in givens + queries: # this symptom will be marginalized out, so skip this step
        continue

    newFactor = Factor(symptoms[i-5], condVars = list(diseasesIndices), inputVars = [i], givenVars = givens, givenVals = givenVals, symptomsInit = True)

    if messageTable is None: # initialize messageTable
        messageTable = newFactor
        continue # go to next iteration/factor

    messageTable.mult_factor_symptom(newFactor)

# combine the diseases tables
for i in range(4, 0, -1):
    newFactor = Factor(diseases[i-1], condVars = [0], inputVars = [i], givenVars = givens, givenVals = givenVals, diseasesInit = True)

    if i in queries + givens:
        remainingFactors.append(newFactor)
        continue

    if messageTable is None: # in the off chance we marginalized away all the symptoms, then the message table will begin as a disease table
        messageTable = newFactor
        continue

    messageTable.mult_factor_disease(newFactor, i)

# combine the isSummer node
for i in [0]:
    newFactor = Factor(summerTable, condVars = [0], inputVars = [], givenVars = givens, givenVals = givenVals, summerInit = True)
    remainingFactors.append(newFactor)

    if messageTable is None:
        messageTable = newFactor # really weird scenario where no variables are given, so they're all marginalized out and all we get is P(x0)
        continue

remainingFactors.append(messageTable)
messageTable.mult_factor_multiple(remainingFactors, 0) # marginalize over 0

messageTable.normalize_table()
print("{} <-- variables in order".format(messageTable.get_vars()))
messageTable.print_table()
print("Sum:", np.sum(messageTable.table))
